op-supervisor/supervisor/backend/backend.go

```diff
@@ -441,9 +439,105 @@ func (su *SupervisorBackend) checkAccess(acc types.Access, execAt types.Executin
+		rpcResultChan <- rpcCheckResult{block: msgBlockRPC, err: err}
+	}()
+
+	msgBlockDB, err := primaryAccessCheck(acc)
+	if err != nil {
+		su.logger.Debug("Access-list inclusion check failed", "err", err)
+		return eth.BlockID{}, err
+	}
+
+	rpcResult := <-rpcResultChan
+	if rpcResult.err != nil {
+		su.logger.Error("RPC access check failed", "err", rpcResult.err)
+		return eth.BlockID{}, rpcResult.err
+	}
+	if msgBlockDB != rpcResult.block {
+		su.logger.Error("RPC access check failed, DB access check result did not match rpc access check result", "db_block", msgBlockDB, "rpc_block", rpcResult.block)
+		return eth.BlockID{}, fmt.Errorf("RPC access check failed, DB access check result did not match rpc access check result")
+	}
+	return rpcResult.block, nil
+}
+
+func (su *SupervisorBackend) checkAccessWithRPC(ctx context.Context, acc types.Access) (eth.BlockID, error) {
+	rpcAccessVerifyTimeout := 5 * time.Second
+
+	src, ok := su.syncSources.Get(acc.ChainID)
+	if !ok {
+		return eth.BlockID{}, fmt.Errorf("%w: %v", types.ErrUnknownChain, acc.ChainID)
+	}
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, rpcAccessVerifyTimeout)
+	defer cancel()
+
+	blockRef, err := src.BlockRefByNumber(timeoutCtx, acc.BlockNumber)
+	if err != nil {
+		return eth.BlockID{}, fmt.Errorf("failed to get block ref for verifiying access with RPC: %w", err)
+	}
+
+	receipts, err := src.FetchReceipts(timeoutCtx, blockRef.Hash)
+	if err != nil {
+		return eth.BlockID{}, fmt.Errorf("failed to fetch receipts for verifying access with RPC: %w", err)
+	}
+
+	targetLogIndex := uint(acc.LogIndex)
+
+	// Find the receipt that might contain our log
+	receiptIndex, found := slices.BinarySearchFunc(receipts, targetLogIndex, func(receipt *ethtypes.Receipt, target uint) int {
+		if len(receipt.Logs) == 0 {
+			return -1 // Continue search
+		}
+
+		if target < receipt.Logs[0].Index {
+			return 1 // Target is smaller, so range comes "after" target in search space
+		}
+		if target > receipt.Logs[len(receipt.Logs)-1].Index {
+			return -1 // Target is larger, so range comes "before" target in search space
+		}
+		return 0 // Target is within the range (inclusive)
+	})
+	if !found {
+		return eth.BlockID{}, fmt.Errorf("Assertion failure, log index %d not found in receipts via binary search", acc.LogIndex)
+	}
+	receipt := receipts[receiptIndex]
+
+	// Find our log in the receipt
+	i, found := slices.BinarySearchFunc(receipt.Logs, targetLogIndex, func(log *ethtypes.Log, target uint) int {
+		return int(log.Index) - int(target)
+	})
+	if !found {
+		return eth.BlockID{}, fmt.Errorf("Assertion failure, log index %d not found in receipt via binary search", acc.LogIndex)
+	}
+	log := receipt.Logs[i]
+	logHash := processors.LogToLogHash(log)
+	entryChecksum := types.ChecksumArgs{
+		BlockNumber: acc.BlockNumber,
+		LogIndex:    acc.LogIndex,
+		ChainID:     acc.ChainID,
+		Timestamp:   blockRef.Time,
+		LogHash:     logHash,
+	}.Checksum()
+	if entryChecksum != acc.Checksum {
```

Comment:
I think we can only compare externally fetched fields like Timestamp and LogHash for optimization.